Introduce the public module `parse`

prantlf opened the pull request:
Attempts to fix #13  by introducing `amodro-trace/parse`.

**`parse: function (contents, options)`**

Exposes `parseFileContents` from `lib/parse`.

Extension: Add method `parseFileContents` to `lib/parse` as a thin wrapper of
`esprima.parse`. Abstract the client from Esprima and saves it from
having a direct dependency on `esprima`.

**`traverse: function (node, visitor)`**

Exposes `traverse` from `lib/parse`.

Extension: Pass parent node to visitors of traverse and traverseBroad too. Help
scenarios, when just knowing the parent is enough and generating the
parent link in every node as an extra parsing step is not needed.

**`findDependencies: function(contents)`**

Exposes `findDependencies` from `lib/parse`.

Extension: Make method `findDependencies` accept an already parsed astRoot too. If
the astRot has been already parsed for other purposes, no need to do it
again.

Extension: Make method `findDependencies` return not only the list of dependent modules, but also the list of formal parameters assigned to exports of the dependent modules, if there are any.

**`findCjsDependencies: function(contents)`**

Exposes `findCjsDependencies` from `lib/parse` with the same extensions as done for `findDependencies` above.

**Note:** In addition to the list of dependent modules, I need also the formal parameter names, to be able to identify, which variable represents the dependent module export in the current module. It is naturally available, when parsing the `define` or `require` function prototypes. I didn't rewrite multiple methods inside `lib/parse`, which return an array of module names, to return an object with two arrays. I just returned the parameter names on the result array as `.params`. Public methods in `amodro-trace/parse` return the modules and parameters as an object: `{modules, params}`.

 I added documentation and unit tests to "whet your appetite" :-)

prantlf committed:
Add module "amodro-trace/parse" as the public interface for "lib/parse"

Add method `parse` to "amodro-trace/parse"
------------------------------------------

Add method `parseFileContents` to "lib/parse" as a thin wrapper of
`esprima.parse`. Saves the client from having a direct dependency on
`esprima`.

Add method `traverse` to "amodro-trace/parse"
---------------------------------------------

Pass parent node to visitors of traverse and traverseBroad too. Help
scenarios, when just knowing the parent is enough and generating the
parent link in every node is not needed.

Add method `findDependencies` to "amodro-trace/parse"
-----------------------------------------------------

Make method `findDependencies` accept an already parsed astRoot too. If
the astRot has been already parsed for other purposes, no need to do it
again.

prantlf committed:
Support the simplified CJS syntax in parse.findDependencies when parsing a file with the single wrapping require() statement too

Only files wrapped by define() worked earlier.

prantlf committed:
Recognize property name as dependency's identifier

Look up the dependent module by `a` in {a: require('a')}.

prantlf committed:
Expose findCjsDependencies from amodro-trace/parse

Use the same input and output parameters as for findDependencies.

Return variable names, which contain results of the require()
statements, in addition to the required module paths.

Reuse the method "findRequireDepNames", which looks for the
dependencies in simple CJS wrappers of AMD modules.

jrburke commented on lib/parse.js:
```diff
@@ -197,6 +207,21 @@ define(['esprima', './lang'], function (esprima, lang) {
+            range: localOptions.range,
+            loc: localOptions.loc
+        });
+    }
```
I like keeping the changes you have for the `parse` methods to take an AST Object or a JS String, and just skips the esprima call if the passed in value is the AST Object. However I wonder if we need this particular method in the API.

I would prefer to not have an API that is basically "call esprima". If it helps, maybe we can expose the version of esprima used by this package as `parse.esprima` if you wanted to make sure to use a version of esprima that matches the AST structure that `parse` works on.

prantlf replied:
Yes, I admit, that saving the `esprima` reference in `vscode-requirejs` and making `amodro-trace` be responsible for parsing too was a little selfish ;-) On the other hand, if `amodro-trace` provides AST traversing and analysis, why not parsing too?

I don't insist on having the `parse` method here. I'm it and the related code. I'll put it to `vscode-requirejs` and use `esprima` too.

jrburke commented on parse.js:
```diff
@@ -0,0 +1,86 @@
+   */
+  parse: function (contents, options) {
+    return privateParse.parseFileContents('', contents, options);
+  },
```
See comment about `parseFileContents`, possible removal.

prantlf replied:
Yes, I'm removing it. And the documentation and tests too.

jrburke commented on lib/parse.js:
```diff
@@ -396,8 +422,10 @@ define(['esprima', './lang'], function (esprima, lang) {
             if (funcArgLength) {
                 deps = (funcArgLength > 1 ? ["require", "exports", "module"] :
                         ["require"]).concat(deps);
+                params = getFormalParameters(node).concat(params);
```
Just a note for me to check this more when I get back to the right computer.

prantlf replied:
It concatenates formal parameter names from the function prototype with variable names from the module callback body. If it is pure AMD, usually only the former array is filled. If it is a CJS-like AMD wrapper, the former array contains just the pseudo-modules and the latter array contains the  rest of dependency names.

jrburke reviewed (changes requested):
In general this is looking really good, and I am interested in getting this into the tree. Unfortunately I will not be with a computer where I can work with this repo until after January 6.

If you wanted to use something like this in the updated VS plugin, [npm allows installing from github](http://remarkablemark.org/blog/2016/09/19/npm-install-from-github/), so you could use your fork until we sort out the final landing.

I did a review by quick diff inspection and mostly had a feedback about the `parse.parse` API. When I get my full computer back, then I will work on getting this integrated into the repo. I also want to look a bit more at the `getFormalParameters` additions.

Thank you for the thoughtful issue/pr and tests, really nice! In the meantime, be sure to sign the [JS Foundation CLA](https://cla.js.foundation/) so these bigger contributions are all clear to land.

prantlf committed:
Remove the method "parse" from the public interface

* Remove documentation unit tests for it. Suggest using an ESTree Spec
compliant parser - Esprima.
* Remove the private method "parseFileContents".

prantlf commented:
Thanks for your interest! I hope, that we improve the change, so that it gets accepted.

No problem with continuing in January. I will use the `public-parse` branch as npm dependency in my fork of `vscode-requirejs`.

I tried the JS Foundation CLA Assistant, but after signing in with my Github account the page just doesn't change. It still appears to wait for a sign-in and doesn't say anything about what it is for. I will read more about the JS Foundation at https://js.foundation/ to understand, why you request it in addition to accepting the license of your project.

Merry Xmas and Happy New Year! :-)

prantlf committed:
Expose `traverseBroad` in `amodro-trace/parse`

* Include `traverseBorad` in public methods exported from this module.
* Fix undeclared identifier in `traverseBroad` in `lib/parse`. JSHint
should really be used regularly...
* Update documentation and unit tests of `traverse` and `traverseBroad`.

prantlf commented:
Eventually I pushed a change to include `traverseBroad` too, to make join the already exposed `traverse`. (I needed to skip blocks with local variable declaration, which shadows an outer variable declaration, which I'm looking up.)

prantlf committed:
Fix crash in parseNode in edge cases

They may not appear usual, but parseNode must not crash with them:

* define({...})
* define([], function (...) {...});
* require({...})
* require([], function (...) {...});

jrburke commented on lib/parse.js:
```diff
@@ -811,18 +833,49 @@ define(['esprima', './lang'], function (esprima, lang) {
+                    arg = node[argPropName][0];
+                    if (arg.type === 'Literal') {
+                        deps.push(arg.value);
+                        if (param) {
```
I am a bit confused by this: `param` is defined outside the `traverse` call, so it could conceivably get a few different values as traverse executes this function. But it is only placed in the `params` if we end up with the current execution finding an `arg.type === 'Literal'`?

Apologies, I have been away from the code for a while, so I may need a refresher. Some tests for `parse.findRequireDepNames` may also help explain/demonstrate what is desired here, or some more comments explaining what the goal is for `param`.

prantlf replied:
`findRequireDepNames` returns the same arrays - `deps` and `params` for CJS modules, as they are (easier) inferred from AMD modules in `recurse`, for example:

```js
define(['dependency'], function (param) {...})
```

```js
deps == ['dependency']
params == ['param']
```

In CJS modules, the visitor in the `traverse` callback uses the `param` variable to track just the previous identifier. Not any previous or parent node across multiple calls to the callback.

```js
var param1 = require('dependency1')
param2 = require('dependency2')
result.param3 = require('dependency2')
```

```js
deps == ['dependency1', 'dependency2', 'dependency3']
params == ['param1', 'param2', 'param3']
```

When the visitor callback of the `traverse` function is called with the identifier `param`, which should be remembered, it will take from one to several calls, until the literal "dependency" in the `require` call will be detected. How to connect it to the `param` with "dependency"?

1. One way would be to stop at the visit of VariableDeclaration, AssignmentExpression or MemberExpression, and do all investigation of left and right expression sides _in the same callback call_. Analyze the left side, sometimes deeper to ObjectExpression and Property, and look at the right side to match the `require`  CallExpression with the module Literal.

2. Another way would be to remember the `param` Identifier, as soon it is detected, and check the the `require`  CallExpression _in another call to the visitor_. Conditions are simpler - there are less "dotted" dereferencing expressions for the current AST node and its related ones in this case, but I needed to stick the module's parameter name to the required module name between two unrelated visitor calls. I did it by remembering the parameter in the outside `param` variable, when encountered, and let it "wait" there. If I hit expected node types, and eventually the `require` expression, I would use the parameter name. If I hit other node type, then expected in subsequential visitor calls, the code didn't follow to the expected expression. I invalidate the parameter name, so that it cannot be connected to a future `require`. just in case. For example:

```js
foo = require('foo')
bar = 42
require('bar')
```

The result should be:

```js
deps == ['foo', 'bar']
params == ['foo']
```

The `bar` dependency has no parameter.

Yes, I could add isolated unit tests for findRequireDepNames. Or more test cases cases for findDependencies and findCjsDependencies, actually. I will look at it.

Do you want me to retain the (current) implementation approach 2, or do you want me to try the approach 1?

jrburke replied:
I think approach 1 will be clearer going forward, and likely also clear up the need for the node.type check below. As it is now, it feels a bit like an implicit state machine and I am not confident enough with the knowledge of AST permutations that it would always result in the desired output.

However, if you like, you can just add some more tests to validate the expected output and we can land as-is. Then I can work on restructuring it if you are running out of time. I want to be sensitive to the amount of time this PR has been open. Totally your call though if you want to see it all the way through to the approach 1.

jrburke commented on lib/parse.js:
```diff
@@ -811,18 +833,49 @@ define(['esprima', './lang'], function (esprima, lang) {
-                if (arg.type === 'Literal') {
-                    deps.push(arg.value);
+                // If range or loc options are set, nodes without type occur.
+                if (node.type) {
```
So this is saying that if the node does have a type that the `param` should be cleared? On first read of the comment above this line, it gave me to impression that `if (!node.type)` was the interesting case?

prantlf replied:
On the first impression, this condition shouldn't be needed. Previous conditions end with `return`. If the code runs to the `param = null`, it should be an unexpected expression and it wouldn't make sense to retain the current parsing state (the parameter name).

It is true with the default Esprima parsing options. However, if `loc` or `range` are enabled, the AST contain nodes without type - whitespace, for example. That is why I need to ignore not only visitor calls, where `node == null`, but also where `node.type == null`. I could've changed the [line 841 in `findRequireDepNames`](https://github.com/amodrojs/amodro-trace/pull/14/files#diff-9f97a9c36fbbfb0d19068d6fae3bb84eR841) from:

```js
if (node) {
```

to:

```js
if (node && node.type) {
```

Not to need that condition and have just `param = null`.

jrburke commented on lib/parse.js:
```diff
@@ -566,45 +579,54 @@ define(['esprima', './lang'], function (esprima, lang) {
      * require('stringLiteral')
+     * @param {String} fileName
+     * @param {String} fileContents (or astRoot)
+     * @param {Object} options
```
This signature does not appear to take an `options` parameter.

prantlf replied:
Ah, my mistake, sorry. `findDependencies` has the options, `findCjsDependencies` not. I'm removing it.

jrburke commented on README.md:
```diff
@@ -517,11 +517,101 @@ var config = require('amodro-config')
+
+var astRoot = esprima.parse(fileContents, {
+  range: includeIndexBasedRangeLocation,
+  loc: includeLineAndColumnBasedLocation
```
`includeIndexBasedRangeLocation` and `includeLineAndColumnBasedLocation` are mentioned in a couple of the comment blocks, but do not show up anywhere except in comment blocks. If we need range and loc set, then we should just use `true` here, or if it does not matter to our code, then the examples can just be `esprima.parse(fileContents)`.

I also think it is

prantlf replied:
I'm removing the options as you suggested. They are not important for the amodro-trace interface.

It was a clumsy way to remind the user by just variable names without comments, that Esprima has options and some of them can help, if they need to connect the AST node with the source code location in an editor, or if they need to generate the source code from AST again. Esprima users will use Esprima documentation anyway.

jrburke reviewed (commented):
Feels like it is close to landing, mostly some nits and making sure I understand one of the new methods completely. See inline comments.

prantlf committed:
Do not try explaining some esprima options in the general code example

Remove options from the examples. They are not important for the
amodro-trace interface. Esprima users will use Esprima documentation.

prantlf committed:
Remove the unused "options" parameter from JSDoc and function calls

findCjsDependencies does not support this parameter and it is optional
for findDependencies.
